lib/irq.h

```diff
@@ -95,6 +143,158 @@ void metal_irq_enable(unsigned int vector);
+ *                     base or IRQ range.
+ * @param[in] irq_num  Number of IRQs required
+ * @return irq number base for success, or negative value for failure.
+ */
+int metal_irq_allocate(int irq_base, int irq_num);
+
+/**
+ * @brief	metal_irq_map_softirq
+ *
+ * Map soft IRQ range with input IRQ range which will trigger the soft
+ * IRQs.
+ *
+ * @param[in] source_irq_base Base of source IRQ range.
+ * @param[in] irq_num  Number of IRQs required
+ * @return base of target soft IRQs for success, or negative value for
+ *         failure.
+ */
+int metal_irq_map_softirq(int source_irq_base, int irq_num);
+
+/**
+ * @brief	metal_irq_controller_set_range
+ *
+ * Set IRQ controller IRQ range
+ *
+ * @param[in] cntr Interrupt controller
+ * @param[in] irq_base Base IRQ id of the range managed by the controller
+ * @param[in] irq_num  number of IRQs managed by the controller
+ */
+static inline
+void metal_irq_controller_set_range(struct metal_irq_controller *cntr,
+				    int irq_base, int irq_num)
+{
+	cntr->irq_base = irq_base;
+	cntr->irq_num = irq_num;
+}
+
+/**
+ * @brief	metal_irq_controller_set_irqs
+ *
+ * Set IRQ controller IRQs array
+ *
+ * @param[in] cntr Interrupt controller
+ * @param[in] irqs IRQs array
+ */
+static inline
+void metal_irq_controller_set_irqs(struct metal_irq_controller *cntr,
+				   struct metal_irq *irqs)
+{
+	cntr->irqs = irqs;
+}
+
+/**
+ * @brief	metal_irq_set_handler
+ *
+ * Set IRQ handler
+ *
+ * @param[in] irq_data interrupt structure for the handler
+ * @param[in] hd interrupt handler function
+ * @param[in] arg interrupt handler argument
+ */
+static inline
+void metal_irq_set_handler(struct metal_irq *irq_data,
+			  metal_irq_handler hd, void *arg)
+{
+	irq_data->hd = hd;
+	irq_data->arg = arg;
+}
+
+/**
+ * @brief	metal_irq_handle
+ *
+ * Call registered IRQ handler
+ *
+ * @param[in] irq_data metal IRQ structure
+ * @param[in] irq IRQ id which will be passed to handler
+ * @return IRQ handler status
+ */
+static inline
+int metal_irq_handle(struct metal_irq *irq_data, int irq)
+{
+	if (irq_data->hd) {
+		return irq_data->hd(irq, irq_data->arg);
+	} else {
+		return METAL_IRQ_NOT_HANDLED;
+	}
+}
+
+/**
+ * @brief	metal_irq_get_controller
+ *
+ * Get IRQ controller which manage the specified IRQ
+ *
+ * @param[in] irq IRQ id
+ * @return metal IRQ controller which manages the specified IRQ for success,
+ *         NULL for failure.
+ */
+struct metal_irq_controller *metal_irq_get_controller(int irq);
+
+/**
+ * @brief	metal_irq_controller_set_pending
+ *
+ * Set IRQ to pending state with the IRQ controller
+ *
+ * @param[in] irq IRQ id
+ * @param[in] cntr IRQ controller
+ */
+static inline
+void metal_irq_controller_set_pending(int irq,
+				      struct metal_irq_controller *cntr)
+{
+	cntr->irq_set_state(cntr, irq, METAL_IRQ_PENDING);
+}
+
+/**
+ * @brief	metal_irq_set_pending
+ *
+ * Set IRQ to pending state
+ *
+ * @param[in] irq IRQ id
+ */
+void metal_irq_set_pending(int irq);
```

Comment:
I think this should be moved to softirq.[ch] since it only is applicable to softirq. This way it does not need to use the call back and HW based interrupt controllers do not need to check for METAL_IRQ_PENDING and return error.

Reply:
Agree, will move this function to softirq.h only.